src/cc/bcc_proc.c

```diff
@@ -148,39 +143,66 @@ int bcc_procutils_each_module(int pid, bcc_procutils_modulecb callback,
     name = buf;
     while (isspace(*name))
       name++;
+    mod.name = name;
     if (!bcc_mapping_is_file_backed(name))
       continue;
 
     if (strstr(name, "/memfd:")) {
-      char *memfd_name = _procutils_memfd_path(pid, inode);
+      char *memfd_name = _procutils_memfd_path(pid, mod.inode);
       if (memfd_name != NULL) {
         strcpy(buf, memfd_name);
         free(memfd_name);
-        name = buf;
+        mod.name = buf;
+        enter_ns = 0;
       }
     }
 
-    if (callback(name, begin, end, (uint64_t)offset, true, payload) < 0)
-      break;
+    if (callback(&mod, enter_ns, payload) < 0)
+      return 0;
   }
 
-  fclose(procmap);
+  return -1;
+}
+
+int bcc_procutils_each_module(int pid, bcc_procutils_modulecb callback,
+                              void *payload) {
+  char procmap_filename[128];
+  FILE *procmap;
+  snprintf(procmap_filename, sizeof(procmap_filename), "/proc/%ld/maps",
+           (long)pid);
+  procmap = fopen(procmap_filename, "r");
+  if (!procmap)
+    return -1;
+
+  if(!_procfs_maps_each_module(procmap, pid, callback, payload))
+    goto done;
 
   // Address mapping for the entire address space maybe in /tmp/perf-<PID>.map
   // This will be used if symbols aren't resolved in an earlier mapping.
   char map_path[4096];
   // Try perf-<PID>.map path with process's mount namespace, chroot and NSPID,
   // in case it is generated by the process itself.
-  if (bcc_perf_map_path(map_path, sizeof(map_path), pid))
-    if (callback(map_path, 0, -1, 0, true, payload) < 0)
-      return 0;
+  mod_info mod;
+  memset(&mod, 0, sizeof(mod_info));
+  if (bcc_perf_map_path(map_path, sizeof(map_path), pid)) {
+    mod.name = map_path;
+    mod.end_addr = -1;
+    if (callback(&mod, 1, payload) < 0)
```

Comment:
Here, `bcc_perf_map_path` may already deal with `/proc/<pid>/root`? So callback function does not need `enter_ns`?

Reply:
I thought this was the case as well, but it's doing readlink on `/proc/PID/root` and prepending the path of the symlink to the perf map path. Some processes on the hosts I'm working on have `/proc/PID/root` symlink resolve to `/`, but the inode is not the same as the actual root inode.

When I tried modifying this line to do `callback(&mod, 0, payload)`, one of the symcache tests in `test_c_api` broke. Think it's necessary to do `enter_ns` here.

Reply:
Okay, the original code indeed entered mount ns.
```
int Context::_each_module(const char *modpath, uint64_t, uint64_t, uint64_t,
                          bool, void *p) {
  Context *ctx = static_cast<Context *>(p);
  // Modules may be reported multiple times if they contain more than one
  // executable region. We are going to parse the ELF on disk anyway, so we
  // don't need these duplicates.
  if (ctx->modules_.insert(modpath).second /*inserted new?*/) {
    ProcMountNSGuard g(ctx->mount_ns_instance_.get());
    bcc_elf_foreach_usdt(modpath, _each_probe, p);
  }
  return 0;
}
```

If you got some time, could you dig a little more to find out why this is needed? It is puzzling.

Reply:
cc @palmtenor who may know the answer as well.

Reply:
This is actually tricky. You don't always want to go with proc root for reading perf map, since they are often only in root /tmp. I remember we try both (with either global PID or NS PID), but I need to take a better look at the code.

Reply:
@palmtenor I think [this code](https://github.com/iovisor/bcc/blob/master/src/cc/bcc_proc.c#L169-L182) is what you are referring to. This PR preserves that functionality.